### PR TITLE
UCP: ODPv2 control configuration

### PR DESCRIPTION
**Pull request by ivankochin**
## What
Add the user control that allow to force ODP registration for certain memory types.

## Why ?
At this moment it seems reasonable to force memory registration through ODP for certain memory types on some systems.

### CODE REVIEW

File: src/ucp/core/ucp_context.c [hunk 1]
```diff
@@ -2044,6 +2060,19 @@ static ucs_status_t ucp_fill_config(ucp_context_h context,
             ((context->config.ext.fence_mode == UCP_FENCE_MODE_AUTO) &&
              (context->config.ext.max_rma_lanes > 1));
 
+    if (context->config.ext.odp_reg_mem_types & UCS_BIT(UCP_ODP_MEM_REG_TYPES_AUTO)) {
+        if (context->config.ext.odp_reg_mem_types != UCS_BIT(UCP_ODP_MEM_REG_TYPES_AUTO)) {
+            ucs_error("If UCP_ODP_MEM_REG_TYPES contains 'auto' then it "
```
Comment (rakhmets) [inline src/ucp/core/ucp_context.c:2065]: If this is an error, then the function should return `UCS_INVALID_PARAM`.

File: src/ucp/core/ucp_context.c [hunk 2]
```diff
@@ -2044,6 +2060,19 @@ static ucs_status_t ucp_fill_config(ucp_context_h context,
...
+                      "shouldn't contain other memory types");
+        }
+    } else {
```
Comment (rakhmets) [inline src/ucp/core/ucp_context.c:2068]: Do we allow an empty configuration parameter to be used?

File: src/ucp/core/ucp_context.c [hunk 3]
```diff
@@ -138,6 +138,13 @@ const size_t ucp_context_est_bcopy_bw[UCS_CPU_VENDOR_LAST] = {
     [UCS_CPU_VENDOR_ZHAOXIN]     = UCP_CPU_EST_BCOPY_BW_DEFAULT
 };
 
+static const char *ucp_odp_reg_mem_types[] = {
```
Comment (gleon99) [inline src/ucp/core/ucp_context.c:141]: Following consistency with other similar places, it's `.._type_names`.

File: src/ucp/core/ucp_context.c [hunk 4]
```diff
@@ -444,6 +451,14 @@ static ucs_config_field_t ucp_context_config_table[] = {
    "directory.",
    ucs_offsetof(ucp_context_config_t, proto_info_dir), UCS_CONFIG_TYPE_STRING},
 
+  {"ODP_MEM_REG_TYPES", "auto",
+   "Memory types that are registered as ODP regions by default"
```
[the remark this answers is not in the document]
Comment (brminich) [reply src/ucp/core/ucp_context.c:455]: lets use  `NONBLOCK` (or `NB` in case the name is too long) instead

File: src/ucp/core/ucp_mm.c [hunk 5]
```diff
@@ -576,6 +576,16 @@ ucp_memh_init_from_parent(ucp_mem_h memh, ucp_md_map_t parent_md_map)
     }
 }
 
+static void
+ucp_memh_check_odp_memory_types(ucp_context_h context,
```
Comment (gleon99) [inline src/ucp/core/ucp_mm.c:580]: Maybe rename it?
I'd expect a `check` function to be a boolean predicate or to `assert()` something, rather than apply config.

File: src/ucp/core/ucp_context.c [hunk 6]
```diff
@@ -444,6 +451,14 @@ static ucs_config_field_t ucp_context_config_table[] = {
...
+   "Either 'auto' for default settings or a combination of:\n"
+    " - 'host'        : Default system memory\n"
```
Comment (gleon99) [inline src/ucp/core/ucp_context.c:457]: Excessive spacing?

File: src/ucp/core/ucp_context.c [hunk 7]
```diff
@@ -138,6 +138,13 @@ const size_t ucp_context_est_bcopy_bw[UCS_CPU_VENDOR_LAST] = {
     [UCS_CPU_VENDOR_ZHAOXIN]     = UCP_CPU_EST_BCOPY_BW_DEFAULT
 };
 
+static const char *ucp_odp_reg_mem_types[] = {
+    [UCP_ODP_MEM_REG_TYPES_AUTO] = "auto",
+    [UCP_ODP_MEM_REG_TYPES_HOST] = "host",
+    [UCP_ODP_MEM_REG_TYPES_CUDA] = "cuda",
+    [UCP_ODP_MEM_REG_TYPES_LAST] = NULL
+};
+
```
Comment (brminich) [inline src/ucp/core/ucp_context.c:147]: imo better add `auto` handling to `ucs_config_sscanf_bitmap`

File: src/ucp/core/ucp_context.c [hunk 8]
```diff
@@ -2044,6 +2060,19 @@ static ucs_status_t ucp_fill_config(ucp_context_h context,
             ((context->config.ext.fence_mode == UCP_FENCE_MODE_AUTO) &&
              (context->config.ext.max_rma_lanes > 1));
 
+    if (context->config.ext.odp_reg_mem_types & UCS_BIT(UCP_ODP_MEM_REG_TYPES_AUTO)) {
+        if (context->config.ext.odp_reg_mem_types != UCS_BIT(UCP_ODP_MEM_REG_TYPES_AUTO)) {
```
Comment (brminich) [inline src/ucp/core/ucp_context.c:2064]: this should be handled on parser level

File: src/ucp/core/ucp_context.h [hunk 9]
```diff
@@ -162,6 +162,8 @@ typedef struct ucp_context_config {
     char                                   *select_distance_md;
     /** Directory to write protocol selection information */
     char                                   *proto_info_dir;
+    /** Memory types that are registered using ODP by default */
```
Comment (brminich) [inline src/ucp/core/ucp_context.h:165]: pls use `nonblock registration` terminology instead of `odp`

File: src/ucp/core/ucp_context.c (hunk 2 above)
Comment (ivankochin) [reply src/ucp/core/ucp_context.c]: As far as I know, yes, we do. But this API isn't finalized yet, so maybe it will be changed in the future.

File: src/ucp/core/ucp_context.c [hunk 10]
```diff
@@ -138,6 +138,13 @@ const size_t ucp_context_est_bcopy_bw[UCS_CPU_VENDOR_LAST] = {
     [UCS_CPU_VENDOR_ZHAOXIN]     = UCP_CPU_EST_BCOPY_BW_DEFAULT
 };
 
+static const char *ucp_reg_nb_mem_type_names[] = {
+    [UCP_MEMTYPE_REG_NB_TYPES_AUTO] = "auto",
+    [UCP_MEMTYPE_REG_NB_TYPES_HOST] = "host",
+    [UCP_MEMTYPE_REG_NB_TYPES_CUDA] = "cuda",
+    [UCP_MEMTYPE_REG_NB_TYPES_LAST] = NULL
+};
```
Comment (brminich) [inline src/ucp/core/ucp_context.c:146]: i think we can reuse `ucs_memory_type_names` (like `MEMTYPE_REG_WHOLE_ALLOC_TYPES` does) and use `UCS_MEMORY_TYPE_LAST` to indicate auto value.

File: src/ucp/core/ucp_context.c (hunk 2 above)
Comment (rakhmets) [reply src/ucp/core/ucp_context.c]: Is it ok that `MEMTYPE_REG_NB_TYPES=auto` is the same as `MEMTYPE_REG_NB_TYPES=`?

Comment (ivankochin) [reply src/ucp/core/ucp_context.c]: At this moment, yes. It is going to be different in the one of future configurations.

File: src/ucp/core/ucp_context.c (hunk 10 above)
Comment (yosefe) [reply src/ucp/core/ucp_context.c:146]: I think we can define special config value for UCS_CONFIG_TYPE_BITMAP case, "all", which would set the bitmap to 0xff..ff.
"all" would mean to use any supported (by UCT) memory type.

Comment (brminich) [reply src/ucp/core/ucp_context.c:146]: sounds good, but we would need to be able to handle `auto` as well, so looks like need a bit for this (maybe use `UNKNOWN`)

Comment (yosefe) [reply src/ucp/core/ucp_context.c:146]: what would be difference between "auto" and "all"?

Comment (brminich) [reply src/ucp/core/ucp_context.c:146]: we can apply different default values on different platforms. E.g. sometimes it will be 0

Comment (yosefe) [reply src/ucp/core/ucp_context.c:146]: sounds weird to have specific platform logic in UCP
anyway I think "auto" is not needed at this point. we may consider adding it later.

File: src/ucp/core/ucp_context.c [hunk 11]
```diff
@@ -444,6 +444,13 @@ static ucs_config_field_t ucp_context_config_table[] = {
    "directory.",
    ucs_offsetof(ucp_context_config_t, proto_info_dir), UCS_CONFIG_TYPE_STRING},
 
+  {"MEMTYPE_REG_NB_TYPES", "auto",
+   "Memory types that perform non-blocking registration by default:\n"
+   "page registration may be deferred until it is accessed by the CPU or a transport.\n"
+   "Allowed memory types: cuda, host.",
```
Comment (ivankochin) [inline src/ucp/core/ucp_context.c:450]: @yosefe @brminich are that "default memory types" correct?

Comment (brminich) [reply src/ucp/core/ucp_context.c:450]: no need to mention those, also `cuda` is not going to be supported

File: src/ucp/core/ucp_context.c [hunk 12]
```diff
@@ -444,6 +444,13 @@ static ucs_config_field_t ucp_context_config_table[] = {
    "directory.",
    ucs_offsetof(ucp_context_config_t, proto_info_dir), UCS_CONFIG_TYPE_STRING},
 
+  {"MEMTYPE_REG_NB_TYPES", "",
+   "Memory types that perform non-blocking registration by default:\n"
+   "page registration may be deferred until it is accessed by the CPU or a transport.\n"
+   "Allowed memory types: cuda, host.",
```
Comment (ivankochin) [inline src/ucp/core/ucp_context.c:450]:
```suggestion
   "page registration may be deferred until it is accessed by the CPU or a transport.",
```

File: src/ucp/core/ucp_context.h [hunk 13]
```diff
@@ -364,6 +366,9 @@ typedef struct ucp_context {
            unsigned               count;
            size_t                 *sizes;
         } am_mpools;
+
+        /* Memory types that perform non-blocking registration by default */
+        uint64_t                  reg_nb_mem_types;
```
Comment (yosefe) [inline src/ucp/core/ucp_context.h:371]: why needed here?
use it through context->config.ext

File: src/ucp/core/ucp_mm.c [hunk 14]
```diff
@@ -586,6 +595,8 @@ ucp_memh_init_uct_reg(ucp_context_h context, ucp_mem_h memh, unsigned uct_flags)
     size_t length              = ucp_memh_length(memh);
     ucs_status_t status;
 
+    ucp_memh_apply_reg_nb_flags(context, mem_type, &uct_flags);
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:598]: do it only when there is a miss in rcache - in ucp_memh_register

File: src/ucp/core/ucp_context.c [hunk 15]
```diff
@@ -444,6 +444,12 @@ static ucs_config_field_t ucp_context_config_table[] = {
    "directory.",
    ucs_offsetof(ucp_context_config_t, proto_info_dir), UCS_CONFIG_TYPE_STRING},
 
+  {"MEMTYPE_REG_NB_TYPES", "",
```
[the remark this answers is not in the document]
Comment (ivankochin) [reply src/ucp/core/ucp_context.c:447]: I did the name by this way to be consistent with `MEMTYPE_REG_WHOLE_ALLOC_TYPES` that has the similar interface. Wouldn't it be better to keep this consistency?

[the remark this answers is not in the document]
Comment (yosefe) [reply src/ucp/core/ucp_context.c:447]: MEMTYPE_REG_WHOLE_ALLOC_TYPES is also a bad name

[the remark this answers is not in the document]
Comment (ivankochin) [reply src/ucp/core/ucp_context.c:447]:
```suggestion
  {"REG_NONBLOCK_MEM_TYPES", "",
```

File: src/ucp/core/ucp_mm.c [hunk 16]
```diff
@@ -366,6 +366,15 @@ void ucp_memh_cleanup(ucp_context_h context, ucp_mem_h memh)
     }
 }
 
+static void
+ucp_memh_apply_reg_nb_flags(ucp_context_h context, ucs_memory_type_t mem_type,
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:370]: maybe no need a separate function?
@brminich WDYT?

Comment (brminich) [reply src/ucp/core/ucp_mm.c:370]: yes, we can avoid that

File: src/ucp/core/ucp_context.c [hunk 17]
```diff
@@ -444,6 +444,12 @@ static ucs_config_field_t ucp_context_config_table[] = {
    "directory.",
    ucs_offsetof(ucp_context_config_t, proto_info_dir), UCS_CONFIG_TYPE_STRING},
 
+  {"REG_NONBLOCK_MEM_TYPES", "",
```
Comment (brminich) [inline src/ucp/core/ucp_context.c:447]: general question, are we going to warn/diag if user sets this value to some memory type and we do not have any MD which supports non-blocking registration for such type?

**Review by yosefe — COMMENTED** [review]
Currently missing the part that avoids registering a memory type that was configured in NONBLOCK mode on MDs that do not support nonblocking mode.

File: src/ucp/core/ucp_mm.c [hunk 18]
```diff
@@ -427,6 +426,13 @@ static ucs_status_t ucp_memh_register(ucp_context_h context, ucp_mem_h memh,
                                      UCT_MD_MEM_REG_FIELD_DMABUF_OFFSET;
         }
 
+        if (context->tl_mds[md_index].attr.reg_nonblock_mem_types &
+            (context->config.ext.reg_nb_mem_types & UCS_BIT(mem_type))) {
+            reg_params.flags = uct_flags | UCT_MD_MEM_FLAG_NONBLOCK;
+        } else {
+            reg_params.flags = uct_flags;
```
Comment (brminich) [inline src/ucp/core/ucp_mm.c:433]: let's add some `ucs_trace` here, or better add ODP related info to the trace on line 450 below.

**Review by brminich — APPROVED** [review]

File: src/ucp/core/ucp_mm.c [hunk 19]
```diff
@@ -427,6 +426,13 @@ static ucs_status_t ucp_memh_register(ucp_context_h context, ucp_mem_h memh,
...
+        }
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:434]: need to register only on MD that supports nonblocking reg (on the memory type)
on MD that doesn't support - we don't register AT ALL

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 20]
```diff
@@ -767,7 +767,9 @@ uct_ib_mlx5_devx_check_odp(uct_ib_mlx5_md_t *md,
         } else {
             md->super.config.odp.max_size = 1ul << 28;
         }
+    }
 
+    if (UCT_IB_MLX5DV_GET(cmd_hca_cap, cap, umr_extended_translation_offset)) {
```
Comment (brminich) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:772]: this needs to be a part of another pr in UCT

File: src/ucp/core/ucp_mm.c [hunk 21]
```diff
@@ -419,21 +420,24 @@ static ucs_status_t ucp_memh_register(ucp_context_h context, ucp_mem_h memh,
     }
 
     ucs_for_each_bit(md_index, md_map) {
+        tl_md = &context->tl_mds[md_index];
+
+        if (context->config.ext.reg_nb_mem_types & UCS_BIT(mem_type)) {
```
Comment (brminich) [inline src/ucp/core/ucp_mm.c:425]: minor: can introduce some aux var and initialize it outside the loop, e.g.
`reg_nb_mtype = context->config.ext.reg_nb_mem_types & UCS_BIT(mem_type)`
and then use it here and on line 468

File: src/ucp/core/ucp_mm.c [hunk 22]
```diff
@@ -461,6 +465,17 @@ static ucs_status_t ucp_memh_register(ucp_context_h context, ucp_mem_h memh,
...
+    if ((context->config.ext.reg_nb_mem_types & UCS_BIT(mem_type)) &&
+        (md_map_registered == 0)) {
+        ucs_log(err_level,
+                "Available MDs cannot perform non-blocking registration "
+                "for %s memory type", ucs_memory_type_names[mem_type]);
+
+        if (!(uct_flags & UCT_MD_MEM_FLAG_HIDE_ERRORS)) {
+            status = UCS_ERR_UNSUPPORTED;
+        }
+    }
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:477]: IMO don't need to consider this an error. it's ok we don't register at all

File: src/ucp/core/ucp_mm.c (hunk 21 above)
Comment (yosefe) [reply src/ucp/core/ucp_mm.c:425]: we could initialize the map of memory domains to register on during ucp context initialization according to nonblocking capabilities

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c (hunk 20 above)
Comment (ivankochin) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:772]: Yes, I just add it here in testing purpose. BTW I think that it won't bring any harm if we merge before this UCT PR with this change.

File: src/ucp/core/ucp_mm.c (hunk 21 above)
Comment (brminich) [reply src/ucp/core/ucp_mm.c:425]: but we need to distinguish the cases when `context->config.ext.reg_nb_mem_types` is not set and when ODP is not supported by MD

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c (hunk 20 above)
Comment (brminich) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:772]: yes, but we need to remove it from this PR

File: src/ucp/core/ucp_mm.c (hunk 22 above)
Comment (ivankochin) [reply src/ucp/core/ucp_mm.c:477]: We have assertion that will fail if we consider this case not an error. https://github.com/openucx/ucx/blob/master/src/ucp/core/ucp_mm.c#L849

Should we remove this assertion too?

BTW what is the case when not-registering memory region is OK? As I understand it will just fail during the later calls, am I correct?

Comment (brminich) [reply src/ucp/core/ucp_mm.c:477]: in theory it will fall-back to some (possibly non-optimal) transport

**Review by brminich — APPROVED** [review]

File: src/ucp/core/ucp_context.c [hunk 23]
```diff
@@ -444,6 +444,12 @@ static ucs_config_field_t ucp_context_config_table[] = {
...
+   "Perform non-blocking memory registration for these memory types:\n"
```
[the remark this answers is not in the document]
Comment (ivankochin) [reply src/ucp/core/ucp_context.c:448]:
```suggestion
   "Perform only non-blocking memory registration for these memory types:\n"
```

**Review by brminich — APPROVED** [review]

**Review by yosefe — APPROVED** [review]
pls squash

**Review by brminich — APPROVED** [review]
